lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,778 @@
+        the entire tree.
+
+        Args:
+            dataset: The name of a dataset that might exist somewhere in the
+                file tree
+
+        Raises:
+            DatasetNotFound: the ARCHIVE tree does not contain a tarball that
+                corresponds to the dataset name
+
+        Returns:
+            Either None if the dataset is not found or the Tarball object
+            representing the dataset that was found.
+        """
+        if dataset in self.datasets:
+            return self.datasets[dataset]
+
+        # If we haven't already discovered the dataset, search for it, and
+        # discover just the controller containing that dataset name.
+        for dir in self.archive_root.iterdir():
+            self.logger.info("Checking controller {}", str(dir))
+            if dir.is_dir():
+                for file in dir.glob("*.tar.xz"):
+                    name = Tarball.stem(file)
+                    self.logger.info("Checking tarball {} [{}]", str(file), name)
+                    if name == dataset:
+                        controller = Controller(
+                            dir,
+                            self.options.INCOMING,
+                            self.options.RESULTS,
+                            self.logger,
+                        )
+                        self.controllers[controller.name] = controller
+                        self.datasets.update(controller.tarballs)
+                        return self.datasets[dataset]
+        raise DatasetNotFound(dataset)
+
+    # These are wrappers for controller and tarball operations which need to be
+    # aware of higher-level constructs in the Pbench artifact tree such as the
+    # ARCHIVE, INCOMING, and RESULTS directory branches. These will manage that
+    # higher level environment surrounding the encapsulated class methods.
+    #
+    # create
+    #   Alternate constructor to create a Tarball object and move an incoming
+    #   tarball and md5 into the proper controller directory.
+    #
+    # unpack
+    #   Unpack the ARCHIVE tarball file into a new directory under the
+    #   controller directory in the INCOMING directory tree.
+    #
+    # uncache
+    #   Remove the unpacked directory tree under INCOMING when no longer needed.
+    #
+    # delete
+    #   Remove the tarball and MD5 file from ARCHIVE after uncaching the
+    #   unpacked directory tree.
+
+    def create(self, controller_name: str, dataset: Path) -> Tarball:
+        """
+        Move a dataset tarball and companion MD5 file into the specified
+        controller directory. The controller directory and links will be
+        created if necessary.
+
+        Args:
+            controller: associated controller name
+            dataset: dataset tarball path
+
+        Returns
+            Tarball object
+        """
+        if dataset in self.datasets:
+            raise DuplicateDataset(dataset)
```

Comment:
What's another exception?

Reply:
If we're presented with a  tarball/dataset name and hash, and we hit a name collision, then I think we need to compare the hashes.  If the hashes match, then, arguably we have no work to do -- we just return an alternate success status (noting the existing duplication rather than reporting the creation -- but the stored data ends up the same in either case); if the hashes don't match, then we return an error noting the conflict, and we allow the caller to decide what to do (e.g., it could request a deletion and then re-issue the creation request).

I don't think it really has much to do with idempotency.  Rather, I think it's more a matter of reporting success relative to policy.  For instance, I believe we have an implicit policy of not replacing data (it has to be deleted and re-created, rather than replaced).  So, if we get a name collision, that would be a failure; however, if the hashes match, then it's as though the operation succeeded, so it's not really a failure; and, so, I think it's worth reporting the alternate success.

However, if there's a name collision and the hashes don't match, then that is a failure, and it will require some action on the part of the caller to resolve.

If we don't compare the hashes, then we cannot tell a duplication from a conflict.  I suppose we could report the duplication as an error, return the existing dataset's hash, and let the caller do the comparison, but that still requires us to fetch the hash of the existing dataset, at which point we might as well perform the comparison, which will allow us to report the alternate success.  (And, of course, if the caller is so inclined, it could respond to the alternate success but reporting a duplication _error_ to its caller/user, if that's what _its_ policy requires.)

Nevertheless, in this particular case, we're looking in what I think could be an unpopulated `FileTree` instance to see if a dataset with this name already exists...and I'm pretty sure that the answer will be "no".  It won't be until after we've created the controller (which populates it...right?), that we'll be able to check for a name collision.  So, I'm not sure that this test is accomplishing enough to warrant keeping it.

The question that I find myself unable to answer is, at what point _would_ we detect the name collision, and how hard would it be to deal with disambiguating duplication from conflict there?